server/routerlicious/packages/services-telemetry/src/lumber.ts

```diff
@@ -0,0 +1,149 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { LumberEventName } from "./lumberEventNames";
+import { LogLevel, LumberType, ITelemetryMetadata, ILumberjackEngine } from "./resources";
+
+// Lumber represents the telemetry data being captured, and it uses a list of
+// ILumberjackEngine to emit the data according to the engine implementation.
+// Lumber should be created through Lumberjack. Additional properties can be added through
+// addProperty(). Once the telemetry event is complete, the user must call either success()
+// or error() on Lumber to emit the data.
+export class Lumber<T extends string = LumberEventName> {
+    private readonly _eventName: T;
+    private readonly _properties: Map<string, any>;
+    private readonly _timestamp: string;
+    private readonly startTime: number;
+    private readonly _type: LumberType;
+    private _metadata: ITelemetryMetadata | undefined;
+    private _latencyInMs: number | undefined;
+    private _successful: boolean | undefined;
+    private _message: string | undefined;
+    private _statusCode: string | undefined;
+    private _exception: Error | undefined;
+    private _logLevel: LogLevel | undefined;
+    private completed: boolean;
+    private readonly engineList: ILumberjackEngine[];
+
+    public get eventName(): T {
+        return this._eventName;
+    }
+
+    public get metadata(): ITelemetryMetadata | undefined {
+        return this._metadata;
+    }
+
+    public get properties(): Map<string, any> {
+        return this._properties;
+    }
+
+    public get type(): LumberType {
+        return this._type;
+    }
+
+    public get timestamp(): string {
+        return this._timestamp;
+    }
+
+    public get latencyInMs(): number | undefined {
+        if (this._type === LumberType.Log) {
+            return undefined;
+        }
+        return this._latencyInMs;
+    }
+
+    public get successful(): boolean | undefined {
+        if (this._type === LumberType.Log) {
+            return undefined;
+        }
+        return this._successful;
+    }
+
+    public get message(): string | undefined {
+        return this._message;
+    }
+
+    public get statusCode(): string | undefined {
+        return this._statusCode;
+    }
+
+    public get exception(): Error | undefined {
+        return this._exception;
+    }
+
+    public get logLevel(): LogLevel | undefined {
+        return this._logLevel;
+    }
+
+    constructor(
+        eventName: T,
+        type: LumberType,
+        engineList: ILumberjackEngine[]) {
+        this._eventName = eventName;
+        this._type = type;
+        this.completed = false;
+        this._properties = new Map<string, any>();
+        this.startTime = Date.now();
+        this._timestamp = new Date(this.startTime).toISOString();
```

Comment:
nit: these defaults could be initialized inline, instead of in the constructor. For example,
```ts
class Lumber {
    public readonly startTime: number = Date.now();
    public readonly timestamp: string = new Date(this.startTime).toISOString();
}

const lumber = new Lumber();

console.log(lumber.timestamp); // 2021-06-23T17:09:59.132Z
console.log(new Date(lumber.timestamp).valueOf()); // 1624468199132
console.log(lumber.startTime); // 1624468199132
```